20240808-custom-limit-email-notifications.md

```diff
@@ -0,0 +1,194 @@
+tags: ["satellite", "emails", "limits", "notifications"]
+---
+
+# Project Limit Notification Emails 
+
+## Essentials
+
+Users are able to set custom egress and storage limits on their project. The goal of this design is to send users with custom limits emails when they hit 80% and 100% usage of their custom limits. This addition ensures that a user is aware that they are approaching their current limits, and gives them the opportunity to increase their limits before exceeding them.
+
+### Header
+
+Date: 2024-07-26 
+
+Owner: mobyvb
+
+Accountable:
+- Console Team
+
+Consulted:
+-
+
+Informed:
+-
+
+### Context
+
+Pro users are currently able to set customized project limits for storage and egress. Right now, the only way for a user to check how close they are to their limits is to go to the project dashboard in the satellite UI. A user could also learn about exceeding their limits when uploads or downloads begin failing, but there is no advance notice in this case.
+
+The purpose of "Project Limit Notification Emails" is to inform users automatically when they are getting close to, or exceeding their custom limits.
+
+### Goals
+
+1. Give users advance notice that they are getting close to their custom limits (80% used)
+2. Inform users when they have exceeded their custom limits (100% used)
+
+### Approach / Design
+
+Things we want to avoid:
+* unintentionally sending multiple emails per "threshold event" (e.g. storage usage on project abc has exceeded 80% of the custom limit)
+* failing to send an email for a "threshold event" (e.g. satellite restart at the same time as threshold event, resulting in dropped communications)
+
+These constraints require that we make some DB updates.
+
+#### Config Updates
+
+A single feature flag should be added on the satellite to enable "limit threshold email notifications". It should default to be turned off, until it is turned on in production.
+
+#### DB Updates
+
+For each project, we need to track whether emails were sent for each "threshold event":
+
+* storage usage exceeded 80% of custom storage limit
+* storage usage exceeded 100% of custom storage limit
+* bandwidth usage exceeded 80% of custom bandwidth limit
+* bandwidth usage exceeded 100% of custom bandwidth limit
+
+We could add a new column to the projects table for each of these, but we can also minimize the schema change by using the bits of a single int field for each one. Example of one approach accomplishing this via a Go struct:
+
+```golang
+type ProjectNotificationFlags struct {
```

Comment:
TODO: design these flags to be more "generic"
They can still be tied to 80/100% thresholds, but the system should be designed in such a way that more customizable limits are possible in the future, without needing to significantly change the code